src/borg/archiver.py

```diff
@@ -696,6 +698,195 @@ def peek_and_store_hardlink_masters(item, matched):
+            # communication with the process is a one-way road, i.e. the process can never block
+            # for us to do something while we block on the process for something different.
+            filterproc = subprocess.Popen(filter, stdin=subprocess.PIPE, stdout=filterout, shell=True)
+            # Always close the pipe, otherwise the filter process would not notice when we are done.
+            tarstream = filterproc.stdin
+            tarstream_close = True
+
+        # The | (pipe) symbol instructs tarfile to use a streaming mode of operation
+        # where it never seeks on the passed fileobj.
+        tar = tarfile.open(fileobj=tarstream, mode='w|')
+
+        self._export_tar(args, archive, tar)
+
+        # This does not close the fileobj (tarstream) we passed to it -- a side effect of the | mode.
+        tar.close()
+
+        if tarstream_close:
+            tarstream.close()
+
+        if filter:
+            logger.debug('Done creating tar, waiting for filter to die...')
+            rc = filterproc.wait()
+            logger.debug('filter exited with code %d', rc)
+            self.exit_code = max(self.exit_code, rc)
+
+            if filterout_close:
+                filterout.close()
+
+        return self.exit_code
+
+    def _export_tar(self, args, archive, tar):
+        matcher = self.build_matcher(args.patterns, args.paths)
+
+        progress = args.progress
+        output_list = args.output_list
+        strip_components = args.strip_components
+        partial_extract = not matcher.empty() or strip_components
+        hardlink_masters = {} if partial_extract else None
+
+        def peek_and_store_hardlink_masters(item, matched):
+            if (partial_extract and not matched and hardlinkable(item.mode) and
+                    item.get('hardlink_master', True) and 'source' not in item):
+                hardlink_masters[item.get('path')] = (item.get('chunks'), None)
+
+        filter = self.build_filter(matcher, peek_and_store_hardlink_masters, strip_components)
+
+        if progress:
+            pi = ProgressIndicatorPercent(msg='%5.1f%% Extracting: %s', step=0.1, msgid='extract')
+            pi.output('Calculating size')
+            extracted_size = sum(item.get_size(hardlink_masters) for item in archive.iter_items(filter))
+            pi.total = extracted_size
+        else:
+            pi = None
+
+        def item_content_stream(item):
+            """
+            Return a file-like object that reads from the chunks of *item*.
+            """
+            chunk_iterator = archive.pipeline.fetch_many([chunk_id for chunk_id, _, _ in item.chunks])
+            if pi:
+                info = [remove_surrogates(item.path)]
+                return ChunkIteratorFileWrapper(chunk_iterator,
+                                                lambda read_bytes: pi.show(increase=len(read_bytes), info=info))
+            else:
+                return ChunkIteratorFileWrapper(chunk_iterator)
```

Comment:
No, extract doesn't use CIFW and recreate generates it's progress display on the other side of it (via chunk_file).